File: platforms/ios/Dezel/Dezel/Core/JavaScriptCallback.swift
/**
 * @class JavaScriptCallback
 * @since 0.1.0
 */
open class JavaScriptCallback: NSObject {

	//--------------------------------------------------------------------------
	// MARK: Properties
	//--------------------------------------------------------------------------

	/**
	 * @property context
	 * @since 0.1.0
	 */
	private(set) public var context: JavaScriptContext

	/**
	 * @property target
	 * @since 0.1.0
	 */
	private(set) public lazy var target: JavaScriptValue = {
		return JavaScriptValue.create(self.context, handle: self.callbackTarget, protect: false)
	}()

	/**
	 * @property callee
	 * @since 0.1.0
	 */
	private(set) public lazy var callee: JavaScriptValue = {
		return JavaScriptValue.create(self.context, handle: self.callbackCallee, protect: false)
	}()

	/**
	 * @property arguments
	 * @since 0.1.0
	 */
	private(set) public var arguments: Int

	/**
	 * @property result
	 * @since 0.1.0
	 * @hidden
	 */
	private(set) internal var result: JSValueRef?

	/**
	 * @property argc
	 * @since 0.1.0
	 * @hidden
	 */
	private(set) internal var argc: Int

	/**
	 * @property argv
	 * @since 0.1.0
	 * @hidden
	 */
	private(set) internal var argv: UnsafePointer<JSValueRef?>

	/**
	 * @property callbackTarget
	 * @since 0.2.0
	 * @hidden
	 */
	private var callbackTarget: JSObjectRef

	/**
	 * @property callbackCallee
	 * @since 0.2.0
	 * @hidden
	 */
	private var callbackCallee: JSObjectRef

	//--------------------------------------------------------------------------
	// MARK: Methods
	//--------------------------------------------------------------------------

	/**
     * @constructor
     * @since 0.1.0
	 * @hidden
     */
	public init(context: JavaScriptContext, target: JSObjectRef, callee: JSObjectRef, argc: Int, argv: UnsafePointer<JSValueRef?>) {

		self.context = context

		self.callbackTarget = target
		self.callbackCallee = callee
		self.arguments = argc

		self.argc = argc
		self.argv = argv
	}

	/**
	 * @method returns
	 * @since 0.1.0
	 */
	final public func returns(_ value: JavaScriptValue?) {
		self.result = toJs(value, in: self.context)
	}

	/**
	 * @method returns
	 * @since 0.1.0
	 */
	final public func returns(_ property: JavaScriptProperty) {
		self.result = toJs(property, in: self.context)
	}

	/**
	 * @method returns
	 * @since 0.1.0
	 */
	final public func returns(_ value: String) {
		self.result = JavaScriptValueCreateString(self.context.handle, value)
	}

	/**
	 * @method returns
	 * @since 0.1.0
	 */
	final public func returns(_ value: Double) {
		self.result = JavaScriptValueCreateNumber(self.context.handle, value)
	}

	/**
	 * @method returns
	 * @since 0.3.0
	 */
	final public func returns(_ value: Float) {
		self.result = JavaScriptValueCreateNumber(self.context.handle, Double(value))
	}

	/**
	 * @method returns
	 * @since 0.3.0
	 */
	final public func returns(_ value: Int) {
		self.result = JavaScriptValueCreateNumber(self.context.handle, Double(value))
	}

	/**
	 * @method returns
	 * @since 0.1.0
	 */
	final public func returns(_ value: Bool) {
		self.result = JavaScriptValueCreateBoolean(self.context.handle, value)
	}
}

File: platforms/android/Dezel/dezel/src/main/java/ca/logaritm/dezel/core/JavaScriptCallback.kt
package ca.logaritm.dezel.core

import ca.logaritm.dezel.core.external.JavaScriptValueExternal

/**
 * @class JavaScriptCallback
 * @since 0.1.0
 */
open class JavaScriptCallback(context: JavaScriptContext, target: Long, callee: Long, argc: Int, argv: LongArray) {

	//--------------------------------------------------------------------------
	// Properties
	//--------------------------------------------------------------------------

	/**
	 * @property context
	 * @since 0.1.0
	 */
	public var context: JavaScriptContext
		private set

	/**
	 * @property target
	 * @since 0.1.0
	 */
	public val target: JavaScriptValue by lazy {
		JavaScriptValue.create(context, this.callbackTarget, false)
	}

	/**
	 * @property callee
	 * @since 0.1.0
	 */
	public val callee: JavaScriptValue by lazy {
		JavaScriptValue.create(context, this.callbackCallee, false)
	}

	/**
	 * @property arguments
	 * @since 0.1.0
	 */
	public var arguments: Int
		private set

	/**
	 * @property argc
	 * @since 0.1.0
	 * @hidden
	 */
	internal var argc: Int
		private set

	/**
	 * @property argv
	 * @since 0.1.0
	 * @hidden
	 */
	internal var argv: LongArray
		private set

	/**
	 * @property result
	 * @since 0.1.0
	 * @hidden
	 */
	internal var result: Long = 0
		private set

	/**
	 * @property callbackTarget
	 * @since 0.2.0
	 * @hidden
	 */
	private var callbackTarget: Long = 0

	/**
	 * @property callbackCallee
	 * @since 0.2.0
	 * @hidden
	 */
	private var callbackCallee: Long = 0

	//--------------------------------------------------------------------------
	// Methods
	//--------------------------------------------------------------------------

	/**
	 * @constructor
	 * @since 0.1.0
	 */
	init {

		this.context = context

		this.callbackTarget = target
		this.callbackCallee = callee
		this.arguments = argc

		this.argc = argc
		this.argv = argv
	}

	/**
	 * @method returns
	 * @since 0.1.0
	 */
	public fun returns(value: JavaScriptValue?) {
		this.result = toJs(value, this.context)
	}

	/**
	 * @method returns
	 * @since 0.1.0
	 */
	public fun returns(property: JavaScriptProperty) {
		this.result = toJs(property, this.context)
	}

	/**
	 * @method returns
	 * @since 0.1.0
	 */
	public fun returns(string: String) {
		this.result = JavaScriptValueExternal.createString(this.context.handle, string)
	}

	/**
	 * @method returns
	 * @since 0.1.0
	 */
	public fun returns(number: Double) {
		this.result = JavaScriptValueExternal.createNumber(this.context.handle, number)
	}

	/**
	 * @method returns
	 * @since 0.3.0
	 */
	public fun returns(number: Float) {
		this.result = JavaScriptValueExternal.createNumber(this.context.handle, number.toDouble())
	}

	/**
	 * @method returns
	 * @since 0.3.0
	 */
	public fun returns(number: Int) {
		this.result = JavaScriptValueExternal.createNumber(this.context.handle, number.toDouble())
	}

	/**
	 * @method returns
	 * @since 0.1.0
	 */
	public fun returns(boolean: Boolean) {
		this.result = JavaScriptValueExternal.createBoolean(this.context.handle, boolean)
	}
}